### PR TITLE
require ACKs and allow PADDING in the Server Stateless Retry packet

### LINKED ISSUE #879: RTT Measurement for HelloRetryRequest
In #817, @marten-seemann asks why the server can't send an ACK with its HelloRetryRequest.

I had assumed that this ACK would be pointless.  Obviously, the server can only acknowledge a single packet (the Initial packet), and this is implicitly acknowledged.  The only information that the packet might now contain is the ACK Delay field now that we removed the timestamps from the default ACK format.  That assumes that the server is delaying its response in any way or that its processing is somehow significant enough to be worth measuring.  I don't think that the server has any reason to delay the generation of a HelloRetryRequest, and the entire design depends on the processing time required to generate it is minimal.

Did I miss something?

### LINKED ISSUE #1067: Retry packet and PADDING
-08 in Section 5.4.2. says:

>   The payload of the Retry packet contains a single STREAM frame on
>   stream 0 with offset 0 containing the server's cryptographic
>   stateless retry material.  It MUST NOT contain any other frames.  The
>   next STREAM frame sent by the server will also start at stream offset
>   0.

Any reason why PADDING is not allowed? (Insert usual stuff about zero-copy and fixed offsets.)

### PR DESCRIPTION
**Pull request by marten-seemann**
Closes #1067. Updates #627.

### CODE REVIEW

**Comment by ekr** [comment]
I'm not in favor of this for the reasons discussed in #879, so we should discuss before merging.

File: draft-ietf-quic-transport.md [hunk 1]
```diff
@@ -627,10 +627,11 @@ any version negotiation that occurred (see {{version-negotiation}}).  The client
 MAY also retain any observed RTT or congestion state that it has accumulated for
 the flow, but other transport state MUST be discarded.
 
-The payload of the Retry packet contains a single STREAM frame
-on stream 0 with offset 0 containing the server's cryptographic stateless retry
-material. It MUST NOT contain any other frames. The next STREAM frame sent by
-the server will also start at stream offset 0.
+The payload of the Retry packet contains a STREAM frame on
```
Comment (nibanks) [inline draft-ietf-quic-transport.md]: I don't think the first two lines here should be changed. Maybe this is just out of date?

**Comment by MikeBishop** [comment]
This PR needs to be re-targeted to cover the new scope of #627, but it's otherwise good.

**Comment by janaiyengar** [comment]
Ugh, sorry @marten-seemann -- I meant to send a patch to your commit, but ended up committing to your branch instead. Take a look, and respond if you think it's fine. I'm happy to merge if you agree.

File: draft-ietf-quic-transport.md [hunk 2]
```diff
@@ -627,10 +627,11 @@ any version negotiation that occurred (see {{version-negotiation}}).  The client
...
 the flow, but other transport state MUST be discarded.
 
-The payload of the Retry packet contains a single STREAM frame
-on stream 0 with offset 0 containing the server's cryptographic stateless retry
-material. It MUST NOT contain any other frames. The next STREAM frame sent by
-the server will also start at stream offset 0.
+The payload of the Retry packet contains at least two frames. It MUST include a
+STREAM frame on stream 0 with offset 0 containing the server's cryptographic
+stateless retry material. It MUST also include an ACK frame to acknowledge the
+client's Initial packet. It MAY additionally include one ore more PADDING
```
Comment (marten-seemann) [inline draft-ietf-quic-transport.md]: s/ore/or

**Comment by marten-seemann** [comment]
LGTM.

**Pull request merged**